aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,254 @@
+            min_gas: 1,
+            max_gas: 1000,
+            target_utilization: 250000000000, // 250 GB
+            read_curve: copy standard_curve,
+            create_curve: copy standard_curve,
+            write_curve: copy standard_curve,
+        };
+        validate_usage_config(&item_config);
+        validate_usage_config(&item_config);
+        move_to(aptos_framework, StorageGasConfig {
+            item_config,
+            byte_config,
+        });
+
+        assert!(
+            !exists<StorageGas>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS)
+        );
+        move_to(aptos_framework, StorageGas {
+            per_item_read: 0,
+            per_item_create: 0,
+            per_item_write: 0,
+            per_byte_read: 0,
+            per_byte_create: 0,
+            per_byte_write: 0,
+        });
+    }
+
+    fun validate_usage_config(config: &UsageGasConfig) {
+        assert!(config.max_gas >= config.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        assert!(config.target_utilization > 0, error::invalid_argument(EZERO_TARGET_UTILIZATION));
+        validate_curve(&config.read_curve);
+        validate_curve(&config.create_curve);
+        validate_curve(&config.write_curve);
+    }
+
+    fun validate_curve(curve: &vector<Point>) {
+        let len = vector::length(curve);
+        assert!(len >= 2, error::invalid_argument(EINVALID_CURVE_LENGTH));
+        assert!(vector::borrow(curve, 0).x == 0, error::invalid_argument(EINVALID_CURVE_START_POINT));
+        assert!(vector::borrow(curve, len - 1).x == BASIS_POINT_BASE, error::invalid_argument(EINVALID_CURVE_END_POINT));
+        let i = 0;
+        while (i < len - 1) {
+            let cur = vector::borrow(curve, i);
+            let next = vector::borrow(curve, i + 1);
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
+    }
+
+    fun calculate_gas(max_usage: u64, current_usage: u64, curve: &vector<Point>, min_gas:u64, max_gas:u64): u64 {
+        let capped_current_usage = if (current_usage > max_usage) max_usage else current_usage;
+        let num_points = vector::length(curve);
+        let current_usage_bps = capped_current_usage * BASIS_POINT_BASE / max_usage;
+        let (i, j) = (0, num_points - 1);
+        while (i < j) {
+            let mid = j - (j - i) / 2;
+            if (current_usage_bps < vector::borrow(curve, mid).x) {
+                j = mid - 1;
+            } else {
+                i = mid;
+            };
+        };
+        let start = vector::borrow(curve, i);
+        if (i == num_points - 1) {
+            min_gas + (max_gas - min_gas) * start.y / BASIS_POINT_BASE
+        } else {
+            let end = vector::borrow(curve, i + 1);
+            min_gas + (max_gas - min_gas) * (start.y + (current_usage_bps - start.x) * (end.y - start.y) / (end.x - start.x)) / BASIS_POINT_BASE
+        }
```

Comment:
shall we do a `curve_project(curve, normalized_x) -> normalized_y` and do `gas = min + (max - min) * normalized_y / MAX` here?